src/netlink_pm_upstream.c

```diff
@@ -42,60 +43,462 @@
+                + MPTCPD_NLA_ALIGN_OPT(token);
+
+        struct l_genl_msg *const msg =
+                l_genl_msg_new_sized(cmd, payload_size);
+
+        bool const appended =
+                append_local_addr_attr(msg, info)
+                && (token == 0
+                    || l_genl_msg_append_attr(
+                            msg,
+                            MPTCP_PM_ATTR_TOKEN,
+                            sizeof(token),  // sizeof(uint32_t)
+                            &token));
+
+        if (!appended) {
+                l_genl_msg_unref(msg);
+
+                return ENOMEM;
+        }
+
+        return l_genl_family_send(pm->family,
+                                  msg,
+                                  mptcpd_family_send_callback,
+                                  (void *) cmd_name, /* user data */
+                                  NULL  /* destroy */) == 0;
+}
+
+// --------------------------------------------------------------
+//          User Space Path Manager Related Functions
+// --------------------------------------------------------------
+static int upstream_announce(struct mptcpd_pm *pm,
+                             struct sockaddr const *addr,
+                             mptcpd_aid_t id,
+                             mptcpd_token_t token)
+{
+        /**
+         * @todo Add support for the optional network interface index
+         *       attribute.
+         */
+        struct addr_info info = {
+                .addr     = addr,
+                .id       = id,
+                .flags    = MPTCP_PM_ADDR_FLAG_SIGNAL,
+                // .ifindex  = ...
+        };
+
+        /**
+         * Set up MPTCP listening socket.
+         *
+         * @todo This should be optional.
+         */
+        if (mptcpd_lm_listen(pm->lm, addr) == 0)
+                return -1;
+
+        return send_add_addr(pm,
+                             MPTCP_PM_CMD_ANNOUNCE,
+                             "announce",
+                             &info,
+                             token);
+}
+
+struct remove_info
+{
+        struct mptcpd_lm *const lm;
+        struct sockaddr const *const sa;
+};
+
+static void upstream_remove_callback(struct l_genl_msg *msg, void *user_data)
+{
+        static char const op[] = "remove_addr";
+
+        mptcpd_family_send_callback(msg, (void *) op);
+
+        /**
+         * @todo The above @c mptcpd_family_send_callback() function
+         *       also calls @c l_genl_msg_get_error().  We could
+         *       refactor but that may not be worth the trouble since
+         *       @c l_genl_msg_get_error() is not an expensive call.
+         */
+        if (l_genl_msg_get_error(msg) == 0) {
+                struct remove_info *info = user_data;
+
+                /**
+                 * Stop listening on MPTCP socket.
+                 *
+                 * @todo This should be optional.
+                 */
+                (void) mptcpd_lm_close(info->lm, info->sa);
```

Comment:
Hopefully the user made sure to use the port returned from `mptcpd_lm_listen()` in address `info->sa` passed to `mptcpd_lm_close()` here otherwise we'll potentially have a resource leak.

Reply:
Hmm. The RFC throws us a slight curve ball here.

It's possible to advertise an address without a port number. In that case it's assumed to be the same port number as the original subflow. But creating a listening port does require a port number, as does removing that listening port.

This is a potential concern with the ephemeral port assignment, depending on how that's added to the mptcpd API. Is `port == 0` overloaded for both "assign a port" and "leave the port number out of the advertisement"? I'll look at the ephemeral port code and comment there.